### PR TITLE
[compiler] optimize dot with single m or n dim

### PR DESCRIPTION
**Pull request by YellowHCH**
* as title

### CODE REVIEW

File: compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp [hunk 1]
```diff
@@ -132,23 +151,107 @@ struct SimplifyDotGeneralToBroadcastMultiply
...
+    } else if (is_single_m_n()) {
+      // m == 1 or n == 1
+      const bool is_single_m = M == 1;
+      auto lhsRank = lhsType.getRank();
+      auto rhsRank = rhsType.getRank();
+
+      // broadcast lhs and rhs.
+      auto lIota = llvm::iota_range<int64_t>(0, lhsRank, false);
+      auto rIota = llvm::iota_range<int64_t>(0, rhsRank, false);
+      llvm::SmallVector<int64_t, 3> lhsBcastDims(lIota.begin(), lIota.end());
```
Comment (qingyunqu) [inline compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp:200]: I found that `lhsBcastDims = llvm::to_vector(llvm::seq<int64_t>(0, lshRank))` is more simple.

Comment (YellowHCH) [reply compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp:200]: > I found that `lhsBcastDims = llvm::to_vector(llvm::seq<int64_t>(0, lshRank))` is more simple.

looks good!

File: compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp [hunk 2]
```diff
@@ -132,23 +151,107 @@ struct SimplifyDotGeneralToBroadcastMultiply
...
+      return lhsType.getDimSize(lhsType.getRank() - 2) == 1 ||
+             rhsType.getDimSize(rhsType.getRank() - 1) == 1;
+    };
+
+    auto resultRank = op.getType().getRank();
+    auto resultShape = op.getType().getShape();
+    auto resultElementTy = op.getType().getElementType();
+    auto has_batch = dimensionNumbers.getLhsBatchingDimensions().size() != 0 ||
+                     dimensionNumbers.getRhsBatchingDimensions().size();
+    auto B = has_batch ? resultShape[0] : 0;
```
Comment (qingyunqu) [inline compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp:171]: coding style: `has_batch` => `hasBatch`

File: compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp [hunk 3]
```diff
@@ -132,23 +151,107 @@ struct SimplifyDotGeneralToBroadcastMultiply
...
+      block.addArgument(blockValArgType, loc);
+      auto *firstArg = block.args_begin();
+      auto *secondArg = std::next(firstArg);
+      // create reduceOp body.
+      {
+        OpBuilder::InsertionGuard guard(rewriter);
+        rewriter.setInsertionPointToStart(&block);
+        auto addOp = rewriter.create<mhlo::AddOp>(loc, *firstArg, *secondArg);
+        rewriter.create<mhlo::ReturnOp>(loc, mlir::ValueRange{addOp});
+      }
```
Comment (qingyunqu) [inline compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp:247]: Should add some lit test and numerical test.

File: compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp [hunk 4]
```diff
@@ -132,23 +151,107 @@ struct SimplifyDotGeneralToBroadcastMultiply
...
+    auto B = has_batch ? resultShape[0] : 0;
+    auto M = lhsType.getDimSize(lhsType.getRank() - 2);
+    auto N = rhsType.getDimSize(rhsType.getRank() - 1);
+    auto LK =
+        lhsType.getDimSize(dimensionNumbers.getLhsContractingDimensions()[0]);
+    auto RK =
+        rhsType.getDimSize(dimensionNumbers.getRhsContractingDimensions()[0]);
+    auto K = std::max(LK, RK);
+
+    if (is_single_k()) {
```
Comment (qingyunqu) [inline compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp:180]: It's not necessary to check k == 1. Just lowering it to mhlo.reduce with reduce rank size == 1, the mhlo.reduce will be simplified to reshape by SimplifyReduceToReshape pattern.

**Comment by YellowHCH** [comment]
CIRERUN

**Comment by YellowHCH** [comment]
CIRERUN

File: compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp [hunk 5]
```diff
@@ -132,22 +151,159 @@ struct SimplifyDotGeneralToBroadcastMultiply
...
+        return false;
+      }
+      // check: k==1
+      return lhsType.getDimSize(dimNums.getLhsContractingDimensions()[0]) ==
+                 1 &&
+             rhsType.getDimSize(dimNums.getRhsContractingDimensions()[0]) == 1;
+    };
+
+    // Try to match the desired condition as k==1. if not, early return.
+    if (auto dotOp =
```
Comment (qingyunqu) [inline compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp:289]: => 
```c++
if constexpr (std::is_same_v<mhlo::DotGeneralOp, OpTy>)
```

File: compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp [hunk 6]
```diff
@@ -33,6 +33,25 @@ using namespace mlir;
...
+                                   Location &loc) {
+  auto constType = RankedTensorType::get({}, elementTy);
+  if (isa<mlir::FloatType>(elementTy)) {
+    auto constAttr = DenseElementsAttr::get(
+        constType,
+        {APFloat::getZero(cast<mlir::FloatType>(elementTy).getFloatSemantics(),
+                          /*negative=*/false)});
+    return rewriter.create<mhlo::ConstantOp>(loc, constType, constAttr);
+  } else if (isa<mlir::IntegerType>(elementTy) &&
+             elementTy.getIntOrFloatBitWidth() != 8) {
```
Comment (qingyunqu) [inline compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp:46]: why `!= 8`?

Comment (YellowHCH) [reply compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp:46]: > why `!= 8`?

Just copied from other where, it's going to remove this.

**Review by qingyunqu — APPROVED** [review]
LGTM

**Pull request merged**